### PR TITLE
Add leaf ordering to gene and sample hierarchical clustering

### PR DESCRIPTION
**Pull request by mzganec**

### CODE REVIEW

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 1]
```diff
@@ -94,32 +94,66 @@
...
       choices:
-        - label: average
-          value: average
         - label: single
           value: single
+        - label: average
+          value: average
         - label: complete
           value: complete
     - name: log2
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Could be just `log`.

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 2]
```diff
@@ -94,32 +94,66 @@
...
       label: Log2
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Could be `Log transform`

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 3]
```diff
@@ -94,32 +94,66 @@
...
+      label: Gene ordering
+      type: basic:string
+      default: knn
+      description: Method used for ordering of genes.
+      choices:
+        - label: None
+          value: none
+        - label: k-NN
+          value: knn
+        - label: Optimal
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Maybe write the full name "Optimal leaf ordering"

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 4]
```diff
@@ -94,32 +94,66 @@
...
+          value: optimal
+        - label: Simulated annealing
+          value: sa
+    - name: n_keep
+      label: Keep permutations
+      type: basic:integer
+      default: 2147483647
+      description: Number of permutations to keep in optimal gene ordering.
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: sample ordering?

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 5]
```diff
@@ -94,32 +94,66 @@
...
+    - name: n_trials
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Visibility of this field could depend on the value of the `ordering` field.

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 6]
```diff
@@ -94,32 +94,66 @@
...
+      choices:
+        - label: None
+          value: none
+        - label: k-NN
+          value: knn
+        - label: Optimal
+          value: optimal
+        - label: Simulated annealing
+          value: sa
+    - name: n_keep
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Visibility of this field could depend on the value of the `ordering` field.

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 7]
```diff
@@ -133,29 +167,36 @@
...
 
       samplehcluster.py \
-        {% for e in exps %}{{e.exp.file}} {% endfor %} \
-        --sampleids {% for e in exps %}{{e|sample_id}} {% endfor %} \
-        {% if genes %}--genes {% for gene in genes %}{{gene}} {% endfor %}{% endif %} \
-        --linkage {{linkage}} \
-        --dstfunc {{distance}} \
-        {%if log2 %}--log2{% endif %} \
-        --output cluster.json
+      --sample_files {% for e in exps %} {{e.exp.file}} {% endfor %} \
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Use indentation (2 spaces).

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 8]
```diff
@@ -133,29 +167,36 @@
...
+      --linkage {{linkage}} \
+      --distance {{distance}} \
+      {% if log2 %} --log2 {% endif %} \
+      {% if normalization != "none" %} --normalization {{normalization}} {% endif %} \
+      --ordering {{ordering}} \
+      --n_keep {{n_keep}} \
+      --n_trials {{n_trials}} \
+      --output cluster.json
+
+      re-checkrc "Sample hierarchical clustering failed."
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Not sure if `re-checkrc` works if there is an empty line before it.

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 9]
```diff
@@ -133,29 +167,36 @@
...
       re-save cluster cluster.json
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: The Pyhton script (samplehcluster.py) could save the cluster.json directly

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml [hunk 10]
```diff
@@ -133,29 +167,36 @@
...
-      re-checkrc "Sample Hierarhlical clustering failed"
       re-progress 1.0
 
 - slug: clustering-hierarchical-genes
```
Comment (mstajdohar) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Some of the above comments also apply here.

File: resolwe_bio/tools/genehcluster.py [hunk 11]
```diff
@@ -3,185 +3,180 @@
 """Hierarchical clustering of genes."""
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+
```
Comment (mstajdohar) [inline resolwe_bio/tools/genehcluster.py]: Why two empty lines?

**Comment by mzganec** [comment]
Depends on [https://github.com/genialis/resolwe-bio/pull/320](https://github.com/genialis/resolwe-bio/pull/320).

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml (hunk 1 above)
Comment (lukaw3d) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Why

Comment (mzganec) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: This fix requires changes also in other places where the process is called. Other visualizations currently also use `log2` as the name of input parameter. Therefore, it seems best to keep this as it is for now.

File: resolwe_bio/processes/clustering/hierarchical_clustering.yml (hunk 9 above)
Comment (mzganec) [inline resolwe_bio/processes/clustering/hierarchical_clustering.yml]: Skipped to avoid problems with large JSON output.

**Comment by mzganec** [comment]
None of the genes or samples are removed in this implementation. If some of the correlations cannot be computed because of degenerate data set NaNs are replaced by zeros as per discussion with @mstajdohar.

**Pull request merged**